code/photometry.py

```diff
@@ -0,0 +1,171 @@
+    )
+
+    # measure sky noise and mean level
+    # suppress warnings about nans in the calculation
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        skymean, skymedian, skynoise = sigma_clipped_stats(
+            bgsubimage, sigma=3.0)
+
+    return subimage, objsrc, nconfsrcs, skymean, skynoise
+
+
+def make_cutouts(ra, dec, infiles, cutout_width, mosaic_pix_scale):
+    """Create cutouts from science and background images.
+
+    Raise an error if a cutout cannot be extracted.
+    """
+    imgfile, skybgfile = infiles
+
+    # extract science image cutout
+    hdulist = fits.open(imgfile)[0]
+    wcs_info = WCS(hdulist)
+    subimage, nodata_param, x1, y1, subimage_wcs = extract_cutout(
+        ra, dec, cutout_width, mosaic_pix_scale, hdulist, wcs_info
+    )
+
+    # if cutout extraction failed, raise an error
+    if nodata_param is True:
+        raise CutoutError("Cutout could not be extracted")
+
+    # extract sky background cutout
+    # if input is same as above, don't redo the cutout, just return
+    if skybgfile == imgfile:
+        return subimage, x1, y1, subimage_wcs, subimage
+
+    # else continue with the extraction
+    bkg_hdulist = fits.open(skybgfile)[0]
+    bgsubimage, bgnodata_param, bgx1, bgy1, bgimage_wcs = extract_cutout(
+        ra, dec, cutout_width, mosaic_pix_scale, bkg_hdulist, wcs_info
+    )
+
+    # if cutout extraction failed, raise an error
+    if bgnodata_param is True:
+        raise CutoutError("Cutout could not be extracted")
+
+    return subimage, x1, y1, subimage_wcs, bgsubimage
+
+
+def run_tractor(subimage, prf, objsrc, skymean, skynoise):
+    # make the tractor image
+    tim = Image(
+        data=subimage,
+        invvar=np.ones_like(subimage) / skynoise**2,
+        psf=PixelizedPSF(prf),
+        wcs=NullWCS(),
+        photocal=NullPhotoCal(),
+        sky=ConstantSky(skymean),
+    )
+
+    # make tractor object combining tractor image and source list
+    tractor = Tractor([tim], objsrc)  # [src]
+
+    # freeze the parameters we don't want tractor fitting
+    tractor.freezeParam("images")  # now fits 2 positions and flux
+    # tractor.freezeAllRecursive()#only fit for flux
+    # tractor.thawPathsTo('brightness')
+
+    # run the tractor optimization (do forced photometry)
+    # Take several linearized least squares steps
+    fit_fail = False
+    try:
+        tr = 0
+        with suppress_stdout():
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", ".*divide by zero.*")
+                # warnings.simplefilter('ignore')
+                for tr in range(20):
+                    dlnp, X, alpha, flux_var = tractor.optimize(variance=True)
+                    # print('dlnp',dlnp)
+                    if dlnp < 1e-3:
+                        break
+
+    # catch exceptions and bad fits
+    except Exception:
```

Comment:
A good rule of thumb is to be as precise as possible with excepting exceptions, e.g. only let the ones through that are in fact expected rather than the most generic `Exception`.

Reply:
Input from someone more familiar with tractor would be helpful. This is a minor improvement over the bare `except` that was here previously, and it's not clear to me what else to expect. I see the [tractor docs](http://thetractor.org/doc/index.html), but no specific documentation on `Tractor.optimize`. 

However, I did put in an `except Exception` in a different place that I know I can be more specific about, so I'll fix that.

Reply:
Good point, I think we should then migrate this how-to to the confluence docs, something along the line:

- By default, avoid try/except, especially with generic `Exception`, add them only for well-understood and specific exceptions only (e.g. the fitter may raise a specific exception for non conversion, then catch that one only and handle the case, but don't catch everything else as that may make bugs more difficult to discover). 

So far, I think this is the first item that can be best addressed from the beginning, during the initial exploratory phase of the notebook/science use case. 

cc @jkrick